dali/pipeline/operators/op_spec.h

```diff
@@ -32,6 +33,65 @@
 
 namespace dali {
 
+namespace detail {
+
+template <typename T, typename S>
+void copy_vector(std::vector<T> &out, const std::vector<S> &in) {
+  out.reserve(in.size());
+  out.clear();
+  for (auto &v : in) {
+    out.emplace_back(v);
+  }
+}
+
+template <typename T>
+void copy_vector(std::vector<T> &out, const std::vector<bool> &in) {
+  out.reserve(in.size());
+  out.clear();
+  for (auto b : in) {
+    out.emplace_back(b);
+  }
+}
+
+/** @brief This overload simply forwrads the reference */
+template <typename T>
+std::vector<T> &&convert_vector(std::vector<T> &&v) {
+  return std::move(v);
+}
+
+/** @brief This overload simply forwrads the reference */
+template <typename T>
+const std::vector<T> &convert_vector(const std::vector<T> &v) {
+  return v;
+}
+
+/** @brief This overload converts elements from v and returns a vector of converted objects */
+template <typename T, typename S>
+std::enable_if_t<!std::is_same<T, S>::value, std::vector<T>>
+convert_vector(const std::vector<S> &v) {
+  std::vector<T> out;
+  copy_vector(out, std::move(v));
+  return out;
+}
+
+}  // namespace detail
+
+template <typename T>
+struct argument_storage {
+  using type = std::conditional_t<
+      std::is_integral<T>::value || std::is_enum<T>::value,
```

Comment:
Not sure if it is any better, but you could do:
```
(std::is_integral<T>::value && !std::is_same<T, bool>::value) || std::is_enum<T>::value
```
and remove the specialization below

Reply:
I don't think it makes much difference. There will be more specializations anyway (TensorLayout).

Reply:
:+1: